### ISSUE TITLE
Spurious warning when a private library is in the deps of a Coq plugin

### ISSUE THREAD

**Issue by rlepigre** [ocaml/dune#7155]
It seems that [this warning](https://github.com/ocaml/dune/blob/d6faeac7e9c35cf9e1f53660b12d90631465bbda/src/dune_rules/coq/coq_rules.ml#L184) triggers when a plugin has a private library in its (transitive) dependencies.

cc @Alizter

**Comment by Alizter** [comment]
@rlepigre yes that is on purpose, how else would you be able to install that plug in? If you can't install the plug in then the vo files cannot be loaded. I might be missing something here tho. Cc @ejgallego

**Comment by ejgallego** [comment]
@rlepigre , indeed that's what Ali said, if you depend on a private plugin, findlib won't find it.

This is due to us relying on the default dune `OCAMLPATH` that points to `_build/install`

It is possible to add support for private plugins to dune in the same way it is done for binaries, the rules then should:

- setup a `_build/$context/.lib` directory with a build-style layout including the required plugins and META files
- extend OCAMLPATH to include `_build/$context/.lib`

**Comment by rlepigre** [comment]
OK, I'm not sure my description was explicit enough, or I am really missing something. Let me try to rephrase: I have a plugin `my_package.my_plugin` that depends on a public library `other_package.foo` that internally relies on a private library `other_package.__private__.internal`. Are we really expecting a warning in that case?

**Comment by ejgallego** [comment]
How can a public library depend on a private library, maybe I'm misundersanding what you mean by private, but if private libraries are not installed then I don't see how `other_package.foo` can ever work.

**Comment by rlepigre** [comment]
I dunno, the package I was playing with is https://opam.ocaml.org/packages/swipl/, which is built and installed by dune. It has a `META` file that looks like this:
```
package "__private__" (
  directory = "__private__"
  package "bindings" (
    directory = "bindings"
    version = "a8c2ea1"
    description = "Ctypes bindings to describe the swipl FFI."
    requires = "ctypes ctypes.stubs"
    archive(byte) = "bindings.cma"
    archive(native) = "bindings.cmxa"
    plugin(byte) = "bindings.cma"
    plugin(native) = "bindings.cmxs"
  )
  version = "a8c2ea1-dirty"
)
package "swipl" (
  directory = "swipl"
  version = "a8c2ea1"
  description = ""
  requires =
  "ctypes ctypes.foreign ppx_deriving.runtime str swipl.__private__.bindings"
  archive(byte) = "swipl.cma"
  archive(native) = "swipl.cmxa"
  plugin(byte) = "swipl.cma"
  plugin(native) = "swipl.cmxs"
```
And my plugin depends on `swipl.swipl`.

**Comment by ejgallego** [comment]
Ok so that's a private lib that is installed, I get confused by private vs non-public.

Indeed I'm not so expect here, I guess indeed there wouldn't be a problem if that package is part of the transitive deps?

I'm not expert on private stuff, I cc @rgrinberg

**Comment by rgrinberg** [comment]
Yeah, that warning seems wrong. A private library that is installed shouldn't be treated differently from a public library in this case. Just to remind everyone, we have 3 types of libraries

* Private + Not attached to a Package
* Private + Attached to a Package
* Public (always attached to a package)

Everything that is attached to a package will be installed along with the package.

**Comment by ejgallego** [comment]
Thanks @rgrinberg , so I guess the right fix is to remove the warning, but make using a private lib fail in the stanza resolution.

**Comment by Alizter** [comment]
Thanks @rgrinberg for explaining. That was not how I understood things when I set that restriction. I think my confusion may have come from the fact that we consider theories private when they don't have a package. Perhaps this makes more sense for Coq, since we cannot hide a Coq library behind an interface like OCaml can.

We will need to write some tests for plugins with private libraries, since I am still unsure how the findlib plugin loading in Coq will interact.
